### ISSUE TITLE
Template Haskell foreign javascript imports are broken?

### ISSUE THREAD

**Issue by mtolly** [ghcjs/ghcjs#364]
Now that https://ghc.haskell.org/trac/ghc/ticket/9703 has been closed in GHC 7.10 I was eager to move some JS import macros over to template Haskell. But, any import involving a JS expression string (not a simple function name) appears to be broken. Not sure if this is a GHCJS or GHC problem but I figured I'd file it here first. I have GHC 7.10.1, GHCJS c1b6239b0289.

All these programs should use `{-# LANGUAGE TemplateHaskell, JavaScriptFFI #-}`.

## These work:

```hs
foreign import javascript unsafe
  "console.log(5);" five :: IO ()

main :: IO ()
main = five
```

```hs
foreign import javascript unsafe
  "5" five :: IO Int

main :: IO ()
main = five >>= print
```

```hs
foreign import javascript unsafe
  "2 + 3" five :: IO Int

main :: IO ()
main = five >>= print
```

## These don't:

```hs
$([d|
  foreign import javascript unsafe
    "console.log(5);" five :: IO ()
  |])

main :: IO ()
main = five
```

```
[1 of 1] Compiling Main             ( app/Main.hs, dist/build/ghctest-exe/ghctest-exe-tmp/Main.js_o )
Linking Template Haskell (ThRunner1)
Linking dist/build/ghctest-exe/ghctest-exe.jsexe (Main)
Running ghctest-exe...
uncaught exception in Haskell main thread: ReferenceError: five is not defined
ReferenceError: five is not defined
    at h$$lx (/Users/mtolly/git/ghctest/dist/build/ghctest-exe/ghctest-exe.jsexe/all.js:36307:3)
    at Object.h$mainLoop (/Users/mtolly/git/ghctest/dist/build/ghctest-exe/ghctest-exe.jsexe/all.js:10459:25)
    at processImmediate [as _immediateCallback] (timers.js:345:15)
```

```hs
$([d|
  foreign import javascript unsafe
    "5" five :: IO Int
  |])

main :: IO ()
main = five >>= print
```

```
[1 of 1] Compiling Main             ( app/Main.hs, dist/build/ghctest-exe/ghctest-exe-tmp/Main.js_o )
Linking Template Haskell (ThRunner1)
Linking dist/build/ghctest-exe/ghctest-exe.jsexe (Main)
Running ghctest-exe...
uncaught exception in Haskell main thread: ReferenceError: five is not defined
ReferenceError: five is not defined
    at h$mainZCMainzimain1_e (/Users/mtolly/git/ghctest/dist/build/ghctest-exe/ghctest-exe.jsexe/all.js:37706:11)
    at Object.h$mainLoop (/Users/mtolly/git/ghctest/dist/build/ghctest-exe/ghctest-exe.jsexe/all.js:10458:25)
    at processImmediate [as _immediateCallback] (timers.js:345:15)
```

So, in these two programs the JS string is completely dropped, and it's as if no string was given so the Haskell name is also a JS global function name.

```hs
$([d|
  foreign import javascript unsafe
    "2 + 3" five :: IO Int
  |])

main :: IO ()
main = five >>= print
```

```
[1 of 1] Compiling Main             ( app/Main.hs, dist/build/ghctest-exe/ghctest-exe-tmp/Main.js_o )
Linking Template Haskell (ThRunner1)

app/Main.hs:4:3:
    "static 2 + 3" is not a valid ccall impent
    When splicing a TH declaration:
      foreign import javascript unsafe "static 2 + 3" five_0 :: GHC.Types.IO GHC.Types.Int
```

Pretty much any more complicated expression than these produces this same error, `not a valid ccall impent`. This message comes from a function [`parseCImport`](https://github.com/ghc/ghc/blob/c6bb2fc50716a2fc540d55ecddbc5c14e94979f7/compiler/parser/RdrHsSyn.hs#L1235), which as per the name is apparently trying to parse a C FFI declaration into `"function"` or `"header.h function"`. So, basically the JS expression imports are not treated correctly at all.

**Comment by luite** [comment]
Ai, might be the built-in `d` quasiquoter still missing something essential. I need to look into this quickly if this is to make it into 7.10.2

**Comment by mtolly** [comment]
Actually my original code where I encountered this did not use the `d` quasiquoter; it manually constructed an `Exp` with the ADT.

**Comment by luite** [comment]
Ok, I found a problem with the code that converts the TH AST to the HsSyn AST, going to submit a fix

**Comment by luite** [comment]
For the sake of completeness, can you add your non-quasiquoter code that also exhibited the problem, so that I can be sure that I haven't missed anything in testing my fix?

**Comment by mtolly** [comment]
Here's a full program making a `ForeignD` by hand, and also a clearer demonstration of the erroneous output.

`ghctest.cabal`:

```
name:                 ghctest
version:              0.1.0.0
build-type:           Simple
cabal-version:        >= 1.10

executable ghctest-exe
  hs-source-dirs:       app
  main-is:              Main.hs
  other-modules:        Decl
  build-depends:        base, ghcjs-base, template-haskell
  default-language:     Haskell2010
  ghc-options:          -ddump-splices
```

`app/Main.hs`:

```hs
{-# LANGUAGE TemplateHaskell, JavaScriptFFI #-}
module Main where

import Decl
import GHCJS.Marshal

$(ffitest)

main :: IO ()
main = do
  js <- toJSRef "abcdefghijklmnoprqstuvwxyz"
  fromJSRef (slice10 js) >>= print
```

`app/Decl.hs`:

```hs
{-# LANGUAGE TemplateHaskell, JavaScriptFFI #-}
module Decl where

import GHCJS.Types
import Language.Haskell.TH

ffitest :: Q [Dec]
ffitest = do
  t <- [t| JSRef String -> JSRef String |]
  let decl = ForeignD $ ImportF JavaScript Unsafe "$1.slice(10)" (mkName "slice10") t
  return [decl]
```

**Output:**

```
╭─mtolly@mbp1440  ~/g/ghctest
╰─$ cabal configure --ghcjs
Resolving dependencies...
Configuring ghctest-0.1.0.0...
╭─mtolly@mbp1440  ~/g/ghctest
╰─$ cabal build
Building ghctest-0.1.0.0...
Preprocessing executable 'ghctest-exe' for ghctest-0.1.0.0...
[1 of 2] Compiling Decl             ( app/Decl.hs, dist/build/ghctest-exe/ghctest-exe-tmp/Decl.js_o )
[2 of 2] Compiling Main             ( app/Main.hs, dist/build/ghctest-exe/ghctest-exe-tmp/Main.js_o )
Linking Template Haskell (Decl,ThRunner1)
app/Main.hs:7:3-9: Splicing declarations
    ffitest
  ======>
    foreign import javascript unsafe "static $1.slice(10) slice10" slice10
      :: ghcjs-prim-0.1.0.0:GHCJS.Prim.JSRef String
         -> ghcjs-prim-0.1.0.0:GHCJS.Prim.JSRef String
Linking dist/build/ghctest-exe/ghctest-exe.jsexe (Decl,Main)
╭─mtolly@mbp1440  ~/g/ghctest
╰─$ cabal run
Preprocessing executable 'ghctest-exe' for ghctest-0.1.0.0...
Linking dist/build/ghctest-exe/ghctest-exe.jsexe (Decl,Main)
Running ghctest-exe...
uncaught exception in Haskell main thread: ReferenceError: slice10 is not defined
ReferenceError: slice10 is not defined
    at h$$nw (/Users/mtolly/git/ghctest/dist/build/ghctest-exe/ghctest-exe.jsexe/all.js:38572:11)
    at Object.h$mainLoop (/Users/mtolly/git/ghctest/dist/build/ghctest-exe/ghctest-exe.jsexe/all.js:10454:25)
    at processImmediate [as _immediateCallback] (timers.js:345:15)
```

Note how `"$1.slice(10)`" turned into the broken `"static $1.slice(10) slice10"`.

**Comment by luite** [comment]
Great, thanks

**Comment by luite** [comment]
Ok, I have a fix that makes all of these work

**Comment by luite** [comment]
See [GHC trac 10638](https://ghc.haskell.org/trac/ghc/ticket/10638) and [Phab D1070](https://phabricator.haskell.org/D1070)

**Comment by luite** [comment]
This has been merged and will be part of 7.10.2 and up
